packages/instruction-plans/src/transaction-plan-executor.ts

```diff
@@ -1,7 +1,228 @@
+ *   executeTransactionMessage: (message) => {
+ *     const transaction = await signTransactionMessageWithSigners(message);
+ *     await sendAndConfirmTransaction(transaction, { commitment: 'confirmed' });
+ *     return { transaction };
+ *   }
+ * });
+ * ```
+ *
+ * @see {@link TransactionPlannerConfig}
+ */
+export function createTransactionPlanExecutor(config: TransactionPlanExecutorConfig): TransactionPlanExecutor {
+    return async (plan, { abortSignal } = {}): Promise<TransactionPlanResult> => {
+        const context: TraverseContext = {
+            ...config,
+            abortSignal: abortSignal,
+            canceled: abortSignal?.aborted ?? false,
+        };
+
+        const cancelHandler = () => {
+            context.canceled = true;
+        };
+        abortSignal?.addEventListener('abort', cancelHandler);
+        const transactionPlanResult = await traverse(plan, context);
+        abortSignal?.removeEventListener('abort', cancelHandler);
+
+        if (context.canceled) {
+            const abortReason = abortSignal?.aborted ? abortSignal.reason : undefined;
+            throw new SolanaError(SOLANA_ERROR__INSTRUCTION_PLANS__FAILED_TO_EXECUTE_TRANSACTION_PLAN, {
+                cause: findErrorFromTransactionPlanResult(transactionPlanResult) ?? abortReason,
+                transactionPlanResult,
+            });
+        }
+
+        return transactionPlanResult;
+    };
+}
+
+type TraverseContext = TransactionPlanExecutorConfig & {
+    abortSignal?: AbortSignal;
+    canceled: boolean;
+};
+
+async function traverse(transactionPlan: TransactionPlan, context: TraverseContext): Promise<TransactionPlanResult> {
+    const kind = transactionPlan.kind;
+    switch (kind) {
+        case 'sequential':
+            return await traverseSequential(transactionPlan, context);
+        case 'parallel':
+            return await traverseParallel(transactionPlan, context);
+        case 'single':
+            return await traverseSingle(transactionPlan, context);
+        default:
+            transactionPlan satisfies never;
+            throw new SolanaError(SOLANA_ERROR__INVARIANT_VIOLATION__INVALID_TRANSACTION_PLAN_KIND, { kind });
+    }
+}
+
+async function traverseSequential(
+    transactionPlan: SequentialTransactionPlan,
+    context: TraverseContext,
+): Promise<TransactionPlanResult> {
+    const results: TransactionPlanResult[] = [];
+
+    for (const subPlan of transactionPlan.plans) {
+        const result = await traverse(subPlan, context);
+        results.push(result);
+    }
+
+    return transactionPlan.divisible
+        ? sequentialTransactionPlanResult(results)
+        : nonDivisibleSequentialTransactionPlanResult(results);
+}
+
+async function traverseParallel(
+    transactionPlan: ParallelTransactionPlan,
+    context: TraverseContext,
+): Promise<TransactionPlanResult> {
+    const results = await concurrentOperations(
+        transactionPlan.plans.map(plan => () => traverse(plan, context)),
```

Comment:
Not saying you have to do this, but just some prior art to look at: `p-limit's` `limitFunction()`. You could wrap `traverse` in `limitFunction()` and then just blast calls into it.

https://github.com/sindresorhus/p-limit/tree/main?tab=readme-ov-file#limitfunctionfn-options-named-export

Reply:
More prior art: https://github.com/sindresorhus/p-map